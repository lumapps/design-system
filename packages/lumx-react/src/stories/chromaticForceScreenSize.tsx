/* eslint-disable */
import React from 'react';
import isChromatic from 'chromatic/isChromatic';

/** Story decorator used to force a minimum screen size when running in chromatic */
export const chromaticForceScreenSize = (story: any) =>
    isChromatic() ? <div style={{ minWidth: 1200, minHeight: 800 }}>{story()}</div> : story();
