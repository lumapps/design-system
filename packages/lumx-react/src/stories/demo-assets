../../../site-demo/static/demo-assets