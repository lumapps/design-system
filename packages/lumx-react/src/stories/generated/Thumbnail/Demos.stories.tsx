/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/thumbnail/Demos' };

export { App as Combined } from './combined';
export { App as Ratios } from './ratios';
export { App as Sizes } from './sizes';
export { App as Variants } from './variants';
