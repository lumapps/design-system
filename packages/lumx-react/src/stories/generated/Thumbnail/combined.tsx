../../../../../site-demo/content/product/components/thumbnail/react/combined.tsx