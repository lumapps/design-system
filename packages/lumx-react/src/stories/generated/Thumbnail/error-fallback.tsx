../../../../../site-demo/content/product/components/thumbnail/react/error-fallback.tsx