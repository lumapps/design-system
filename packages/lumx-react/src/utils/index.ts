/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
