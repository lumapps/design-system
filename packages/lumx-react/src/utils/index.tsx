export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
