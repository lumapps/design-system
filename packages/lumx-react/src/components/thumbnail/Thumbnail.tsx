import React, {
    forwardRef,
    ImgHTMLAttributes,
    KeyboardEventHandler,
    MouseEventHandler,
    ReactElement,
    ReactNode,
    Ref,
    useRef,
    useState,
} from 'react';
import classNames from 'classnames';

import { AspectRatio, HorizontalAlignment, Icon, Size, Theme } from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { mdiImageBrokenVariant } from '@lumx/icons';
import { isInternetExplorer } from '@lumx/react/utils/isInternetExplorer';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';
import { useFocusPoint } from '@lumx/react/components/thumbnail/useFocusPoint';
import { useImageLoad } from '@lumx/react/components/thumbnail/useImageLoad';
import { useClickable } from '@lumx/react/components/thumbnail/useClickable';
import { FocusPoint, ThumbnailSize, ThumbnailVariant } from './types';

type ImgHTMLProps = ImgHTMLAttributes<HTMLImageElement>;

/**
 * Defines the props of the component.
 */
export interface ThumbnailProps extends GenericProps {
    /** Alignment of the thumbnail in it's parent (requires flex parent). */
    align?: HorizontalAlignment;
    /** Image alternative text. */
    alt: string;
    /** Image aspect ratio. */
    aspectRatio?: AspectRatio;
    /** Badge. */
    badge?: ReactElement;
    /** Image cross origin resource policy. */
    crossOrigin?: ImgHTMLProps['crossOrigin'];
    /** Fallback icon (SVG path) or react node when image fails to load. */
    fallback?: string | ReactNode;
    /** Whether the thumbnail should fill it's parent size (requires flex parent) or not. */
    fillHeight?: boolean;
    /** Apply relative vertical and horizontal shift (from -1 to 1) on the image position inside the thumbnail. */
    focusPoint?: FocusPoint;
    /** Image URL. */
    image: string;
    /** Props to inject into the native <img> element. */
    imgProps?: ImgHTMLProps;
    /** Reference to the native <img> element. */
    imgRef?: Ref<HTMLImageElement>;
    /** Size variant of the component. */
    size?: ThumbnailSize;
    /** Image loading mode. */
    loading?: ImgHTMLProps['loading'];
    /** On click callback. */
    onClick?: MouseEventHandler<HTMLDivElement>;
    /** On key press callback. */
    onKeyPress?: KeyboardEventHandler<HTMLDivElement>;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Variant of the component. */
    variant?: ThumbnailVariant;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Thumbnail';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ThumbnailProps> = {
    fallback: mdiImageBrokenVariant,
    loading: 'lazy',
    theme: Theme.light,
};

/**
 * Thumbnail component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Thumbnail: Comp<ThumbnailProps> = forwardRef((props, ref) => {
    const {
        align,
        alt,
        aspectRatio,
        badge,
        className,
        crossOrigin,
        fallback,
        fillHeight,
        focusPoint,
        image,
        imgProps,
        imgRef: propImgRef,
        loading,
        size,
        theme,
        variant,
        ...forwardedProps
    } = props;
    const imgRef = useRef<HTMLImageElement>(null);

    // Image loading state.
    const loadingState = useImageLoad(imgRef);
    const hasError = loadingState === 'hasError';
    const isLoading = loadingState === 'isLoading';

    const [wrapper, setWrapper] = useState<HTMLElement>();
    const wrapperProps: any = {
        ...forwardedProps,
        ref: mergeRefs(setWrapper, ref),
        className: classNames(
            className,
            handleBasicClasses({ align, aspectRatio, prefix: CLASSNAME, size, theme, variant, hasBadge: !!badge }),
            isLoading && wrapper?.getBoundingClientRect()?.height && 'lumx-color-background-dark-L6',
            fillHeight && `${CLASSNAME}--fill-height`,
        ),
        // Handle clickable Thumbnail a11y.
        ...useClickable(props),
    };

    // Update img style according to focus point and aspect ratio.
    const style = useFocusPoint({ image, focusPoint, aspectRatio, imgRef, loadingState, wrapper });

    return (
        <div {...wrapperProps}>
            <div
                className={`${CLASSNAME}__background`}
                style={{
                    ...style?.wrapper,
                    // Remove from layout if image not loaded correctly (use fallback)
                    display: hasError ? 'none' : undefined,
                    // Hide while loading.
                    visibility: isLoading ? 'hidden' : undefined,
                }}
            >
                <img
                    {...imgProps}
                    style={{
                        ...imgProps?.style,
                        ...style?.image,
                    }}
                    ref={mergeRefs(imgRef, propImgRef)}
                    className={style?.image ? `${CLASSNAME}__focused-image` : `${CLASSNAME}__image`}
                    crossOrigin={crossOrigin && !isInternetExplorer() ? crossOrigin : undefined}
                    src={image}
                    alt={alt}
                    loading={loading}
                />
            </div>
            {hasError &&
                (typeof fallback === 'string' ? (
                    <Icon className={`${CLASSNAME}__fallback`} icon={fallback} size={size || Size.m} theme={theme} />
                ) : (
                    <div className={`${CLASSNAME}__fallback`}>{fallback}</div>
                ))}
            {badge &&
                React.cloneElement(badge, { className: classNames(`${CLASSNAME}__badge`, badge.props.className) })}
        </div>
    );
});
Thumbnail.displayName = COMPONENT_NAME;
Thumbnail.className = CLASSNAME;
Thumbnail.defaultProps = DEFAULT_PROPS;
