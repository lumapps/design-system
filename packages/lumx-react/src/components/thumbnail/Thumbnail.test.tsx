import React from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { commonTestsSuite, itShouldRenderStories } from '@lumx/react/testing/utils';

import { Thumbnail, ThumbnailProps } from './Thumbnail';
import { Clickable, CustomFallback, Default, DefaultFallback, IconFallback, WithBadge } from './Thumbnail.stories';

const CLASSNAME = Thumbnail.className as string;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (props: Partial<ThumbnailProps> = {}, shallowRendering = true) => {
    const renderer: any = shallowRendering ? shallow : mount;
    const wrapper: any = renderer(<Thumbnail {...(props as any)} />);
    return { props, wrapper };
};

describe(`<${Thumbnail.displayName}>`, () => {
    // 1. Test render via snapshot.
    describe('Snapshots and structure', () => {
        itShouldRenderStories(
            { Default, Clickable, DefaultFallback, CustomFallback, IconFallback, WithBadge },
            Thumbnail,
        );
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
