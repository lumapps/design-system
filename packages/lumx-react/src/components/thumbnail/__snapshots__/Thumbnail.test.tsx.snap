// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Thumbnail> Snapshots and structure should render story 'Clickable' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--size-xxl lumx-thumbnail--theme-light"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="Click me"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="landscape1.jpg"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Thumbnail> Snapshots and structure should render story 'CustomFallback' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--theme-light"
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="foo"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="foo"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Thumbnail> Snapshots and structure should render story 'Default' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--size-xxl lumx-thumbnail--theme-light"
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="Image alt text"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="landscape1.jpg"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Thumbnail> Snapshots and structure should render story 'DefaultFallback' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--theme-light"
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="foo"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="foo"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Thumbnail> Snapshots and structure should render story 'IconFallback' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--theme-light"
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="foo"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="foo"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Thumbnail> Snapshots and structure should render story 'WithBadge' 1`] = `
<div
  className="lumx-thumbnail lumx-thumbnail--aspect-ratio-square lumx-thumbnail--size-l lumx-thumbnail--theme-light lumx-thumbnail--variant-rounded lumx-thumbnail--has-badge"
>
  <div
    className="lumx-thumbnail__background"
    style={
      Object {
        "display": undefined,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="Image alt text"
      className="lumx-thumbnail__image"
      loading="lazy"
      src="landscape1.jpg"
      style={Object {}}
    />
  </div>
  <Badge
    className="lumx-thumbnail__badge"
    color="primary"
  >
    <Icon
      icon="M4 2A2 2 0 0 0 2 4V12H4V8H6V12H8V4A2 2 0 0 0 6 2H4M4 4H6V6H4M22 15.5V14A2 2 0 0 0 20 12H16V22H20A2 2 0 0 0 22 20V18.5A1.54 1.54 0 0 0 20.5 17A1.54 1.54 0 0 0 22 15.5M20 20H18V18H20V20M20 16H18V14H20M5.79 21.61L4.21 20.39L18.21 2.39L19.79 3.61Z"
    />
  </Badge>
</div>
`;
